File: step_prototype_new.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import time as time_module
import pickle
import mylib
from scipy.signal.windows import hann
from control import frequency_response
from typing import Optional

# =========================================================================
# 1. DEFINE FILE PATH
# =========================================================================

file_path = "./20250908/20250908_flipmini_00.bbl.csv"

# =========================================================================
# 2. READ THE ENTIRE FILE
# =========================================================================
header_line: int = 0  # Line number where data starts
print("Reading CSV file...")
start_time = time_module.time()

# Read CSV file - pandas will automatically handle the header
# First, we need to find where the actual data starts
with open(file_path, "r") as f:
    line_count = 0
    for line in f:
        line_count += 1
        # Look for loopIteration with or without quotes
        if "loopIteration" in line:
            header_line = line_count - 1  # 0-indexed for pandas
            break

# Load or read data
mat_file_path = file_path.replace(".bbl.csv", ".pkl")

try:
    # Try to load preprocessed data
    with open(mat_file_path, "rb") as f:
        data_dict = pickle.load(f)
        data = data_dict["data"]
        column_names = data_dict["columns"]
    print(f"Loaded preprocessed data from {mat_file_path}")
except FileNotFoundError:
    # Read the CSV file
    df = pd.read_csv(file_path, skiprows=header_line)

    # Store column names and convert to numpy array
    column_names = df.columns.tolist()
    data = df.values

    # Save for future use
    with open(mat_file_path, "wb") as f:
        pickle.dump({"data": data, "columns": column_names}, f)
    print(f"Saved preprocessed data to {mat_file_path}")

elapsed_time = time_module.time() - start_time
print(f"Data loaded: {data.shape[0]} rows in {elapsed_time:.2f} seconds")


def find_column(col_name: str) -> Optional[int]:
    """Find column index by name"""
    try:
        return column_names.index(col_name)
    except ValueError:
        return None


col_time = find_column("time")
col_setpoint_roll = find_column("setpoint[0]")
col_gyro_unfilt_roll = find_column("gyroUnfilt[0]")
col_debug_chirp = find_column("debug[0]")

# =========================================================================
# 3. EXTRACT THE DATA
# =========================================================================

# Time vector (convert from microseconds to seconds)
time = (data[:, col_time] - data[0, col_time]) * 1.0e-6

# Extract chirp signal
chirp = data[:, col_debug_chirp]

# Detect chirp start and end (transition from 0 to non-zero)
chirp_binary = (chirp > 0).astype(int)
chirp_diff = np.diff(np.concatenate([[0], chirp_binary]))  # prepend 0 to align indices
chirp_start_idx = np.where(chirp_diff == 1)[0]  # rising edge
chirp_end_idx = np.where(chirp_diff == -1)[0]  # falling edge

chirp_start = chirp_start_idx[0]
chirp_end = chirp_end_idx[0]


print(chirp_start, chirp_end)

# extract raw data
setpoint_roll = data[:, col_setpoint_roll]
gyro_unfilt_roll = data[:, col_gyro_unfilt_roll]

# Extract data over first chirp
time_over_chirp = time[chirp_start:chirp_end]
setpoint_over_chirp = setpoint_roll[chirp_start:chirp_end]
gyro_over_chirp = gyro_unfilt_roll[chirp_start:chirp_end]

print(f"\nExtracted {len(time)} samples ({time[-1]:.2f} seconds of data)")

# =========================================================================
# 4. PROCESS DATA
# =========================================================================
# Calculate sampling time
Ts = 125 * 1.0e-6
Ts_cntr = 2 * Ts
Ts_log = 2 * Ts_cntr

# Parameters
fs = 1 / Ts
f_max_hz = 650
Nest = round(2.0 / Ts_log)
koverlap = 0.9
Noverlap = round(koverlap * Nest)
window = hann(Nest)

G, C, freq, _ = mylib.estimate_frequency_response(
    setpoint_over_chirp, gyro_over_chirp, window, Noverlap, Nest, Ts_log
)
step_resp = mylib.calculate_step_response_from_frd(G, f_max_hz)

# =========================================================================
# 5. PLOT DATA
# =========================================================================
# PLOT 1
fig1, (ax1, ax2) = plt.subplots(2, 1)

ax1.plot(time, setpoint_roll)
ax1.set_title("setpoint")

ax2.plot(time, gyro_unfilt_roll)
ax2.set_title("gyro unfiltered")

fig1.suptitle("raw data")

# PLOT 2
fig2, (ax1, ax2) = plt.subplots(2, 1)

ax1.plot(time_over_chirp, setpoint_over_chirp)
ax1.set_title("setpoint")

ax2.plot(time_over_chirp, gyro_over_chirp)
ax2.set_title("gyro unfiltered")

fig2.suptitle("data over chirp")

# PLOT 3
G_squeezed = np.squeeze(G.fresp)
plt.figure()
plt.plot(freq, 20 * np.log10(G_squeezed))
plt.xlim(0, f_max_hz)
# plt.ylim(-10.6, 1.2)
plt.suptitle("Frequency Response")

# PLOT 4
plt.figure()
plt.plot(step_resp)
plt.xlim(0, f_max_hz)
plt.suptitle("Frequency Response")

plt.tight_layout()
plt.show()

File: step_prototype_new.m
% =========================================================================
% Simple CSV Reader - Extract Setpoint and GyroUnfilt Data
% No external functions required
% =========================================================================

clc, clear variables
addpath('./lib');
%% 1. DEFINE FILE PATH
file_path = './20250908/20250908_flipmini_00.bbl.csv'

%% 2. READ THE ENTIRE FILE:
fprintf('Reading CSV file...\n');
tic

% Open file
fid = fopen(file_path, 'r');

% Skip header lines - find where data starts
% Header ends where we see a line starting with a number (loopIteration)
line_count = 0;
while ~feof(fid)
    line = fgetl(fid);
    line_count = line_count + 1;
    
    % Check if line starts with "loopIteration" - this is the column header
    if startsWith(line, '"loopIteration"')
        % Parse column names
        column_names = strsplit(line, ',');
        break;
    end
end

% Find the column indices we need
col_time = find(strcmp(column_names, '"time"'));
col_setpoint_roll = find(strcmp(column_names, '"setpoint[0]"'));
col_gyro_unfilt_roll = find(strcmp(column_names, '"gyroUnfilt[0]"'));
col_debug_chirp = find(strcmp(column_names, '"debug[0]"'));

fclose(fid);

% Read data matrix starting after header
try
    load([file_path(1:end-8), '.mat'])
catch exception
    data = readmatrix(file_path, 'NumHeaderLines', line_count);
    save([file_path(1:end-8), '.mat'], 'data')
end

toc
fprintf('Data loaded: %d rows\n', size(data, 1));

%% 3. EXTRACT THE DATA

% Time vector (convert from microseconds to seconds)
time = (data(:, col_time) - data(1, col_time)) * 1.0e-6;

% Extract chirp signal
chirp = data(:, col_debug_chirp);

% Detect chirp start and end (transition from 0 to non-zero)
chirp_binary = chirp > 0;
chirp_diff = diff([0; chirp_binary]);  % prepend 0 to align indices
chirp_start_idx = find(chirp_diff == 1);  % rising edge
chirp_end_idx = find(chirp_diff == -1);   % falling edge

% Get first chirp signal
chirp_start = chirp_start_idx(1);
chirp_end = chirp_end_idx(1);

setpoint_roll = data(:, col_setpoint_roll);
gyro_unfilt_roll = data(:, col_gyro_unfilt_roll);

% Get data over chirp
time_over_chirp = time(chirp_start:chirp_end);
setpoint_over_chirp = setpoint_roll(chirp_start:chirp_end);
gyro_over_chirp = gyro_unfilt_roll(chirp_start:chirp_end);

fprintf('\nExtracted %d samples (%.2f seconds of data)\n', length(time), time(end));

%% 4. PROCESS DATA
% Calculate sampling time
Ts = 125 * 1.e-6;
Ts_cntr = 2 * Ts;
Ts_log = 2 * Ts_cntr;

% Parameters
Nest     = round(2 / Ts_log);
koverlap = 0.9;
Noverlap = floor(koverlap * Nest);
window   = hann(Nest, 'periodic');

[G, C, ~, ~] = estimate_frequency_response(setpoint_over_chirp, gyro_over_chirp, window, Noverlap, Nest, Ts_log);

f_max_hz = 200
step_resp = calculate_step_response_from_frd(G, f_max_hz);

%% 5. PLOT DATA 
figure(1)
subplot(211)
plot(time, setpoint_roll)
title("setpoint")
subplot(212)
plot(time, gyro_unfilt_roll)
title("gyro")

figure(2)
subplot(211)
plot(time_over_chirp, setpoint_over_chirp)
title("setpoint")
subplot(212)
plot(time_over_chirp, gyro_over_chirp)
title("gyro")

figure(3)
plot(G, C);
xlabel("Frequency (Hz)");
title("Frequency Response");
xlim([0 , f_max_hz])
grid on;

figure(4)
plot(step_resp);
title("Step Response")
xlabel("time (ms)");
xlim([0 , f_max_hz])
grid on;
fprintf("\n=== Processing Complete ===\n");
